pkg/client/cloudevents/grpcsource/watcherstore.go

```diff
@@ -28,68 +33,87 @@ import (
 type RESTFulAPIWatcherStore struct {
 	sync.RWMutex
 
-	result         chan watch.Event
-	done           chan struct{}
-	watcherStopped bool
-
 	sourceID  string
 	apiClient *openapi.APIClient
+
+	watchers  map[string]*workWatcher
+	workQueue cache.Queue
 }
 
 var _ store.WorkClientWatcherStore = &RESTFulAPIWatcherStore{}
 
-func NewRESTFullAPIWatcherStore(apiClient *openapi.APIClient, sourceID string) *RESTFulAPIWatcherStore {
-	return &RESTFulAPIWatcherStore{
-		result:         make(chan watch.Event),
-		done:           make(chan struct{}),
-		watcherStopped: false,
-
+func newRESTFulAPIWatcherStore(ctx context.Context, apiClient *openapi.APIClient, sourceID string) *RESTFulAPIWatcherStore {
+	s := &RESTFulAPIWatcherStore{
 		sourceID:  sourceID,
 		apiClient: apiClient,
+		watchers:  make(map[string]*workWatcher),
+		workQueue: cache.NewFIFO(func(obj interface{}) (string, error) {
+			work, ok := obj.(*workv1.ManifestWork)
+			if !ok {
+				return "", fmt.Errorf("unknown object type %T", obj)
+			}
+
+			// ensure there is only one object in the queue for a work
+			return string(work.UID), nil
+		}),
 	}
-}
 
-// ResultChan implements watch interface.
-func (m *RESTFulAPIWatcherStore) ResultChan() <-chan watch.Event {
-	return m.result
+	// start a goroutine to send works to the watcher
+	go wait.Until(s.process, time.Second, ctx.Done())
+
+	return s
 }
 
-// Stop implements watch interface.
-func (m *RESTFulAPIWatcherStore) Stop() {
-	// Call Close() exactly once by locking and setting a flag.
-	m.Lock()
-	defer m.Unlock()
-	// closing a closed channel always panics, therefore check before closing
-	select {
-	case <-m.done:
-		close(m.result)
-	default:
-		m.watcherStopped = true
-		close(m.done)
+// GetWatcher returns a watcher to the source work client with a specified namespace (consumer name).
+// Using `metav1.NamespaceAll` to specify all namespaces.
+func (m *RESTFulAPIWatcherStore) GetWatcher(namespace string, opts metav1.ListOptions) (watch.Interface, error) {
+	// Only list works from maestro server with the given namespace when a watcher is required
+	search := []string{fmt.Sprintf("source = '%s'", m.sourceID)}
+	if namespace != metav1.NamespaceAll {
+		search = append(search, fmt.Sprintf("consumer_name = '%s'", namespace))
+	}
+
+	rbs, _, err := m.apiClient.DefaultApi.ApiMaestroV1ResourceBundlesGet(context.Background()).
+		Search(strings.Join(search, " and ")).
+		Page(1).
+		Size(-1).
+		Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	// save the works to a queue
+	for _, rb := range rbs.Items {
+		work, err := ToManifestWork(&rb)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := m.workQueue.Add(work); err != nil {
```

Comment:
Please see https://github.com/openshift-online/maestro/pull/133#issuecomment-2182223111

Reply:
@machi1990 thanks! updated, we should register the watcher firstly, then add the work to queue